concurrent/actor/EventLoop.java

```diff
@@ -0,0 +1,199 @@
+ */
+
+package grakn.common.concurrent.actor;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.PriorityQueue;
+import java.util.Random;
+import java.util.concurrent.LinkedTransferQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TransferQueue;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+public class EventLoop {
+    private static final Logger LOG = LoggerFactory.getLogger(EventLoop.class);
+    private enum State { READY, RUNNING, STOPPED }
+
+    private State state;
+    private final TransferQueue<Job> jobs = new LinkedTransferQueue<>();
+    private final ScheduledJobQueue scheduledJobs = new ScheduledJobQueue();
+    private final Thread thread;
+
+    public EventLoop(ThreadFactory factory) {
+        thread = factory.newThread(this::loop);
+        state = State.READY;
+        thread.start();
+    }
+
+    public void submit(Runnable job, Consumer<Exception> errorHandler) {
+        jobs.offer(new Job(job, errorHandler));
+    }
+
+    public EventLoop.Cancellable submit(long scheduleMs, Runnable job, Consumer<Exception> errorHandler) {
+        final AtomicReference<Cancellable> cancellable = new AtomicReference<>();
+        submit(() -> cancellable.set(scheduledJobs.offer(scheduleMs, new Job(job, errorHandler))), errorHandler);
+        return cancellable.get();
+    }
+
+    public void await() throws InterruptedException {
+        thread.join();
+    }
+
+    public void stop() throws InterruptedException {
+        submit(
+                () -> state = State.STOPPED,
+                e -> LOG.error("An unexpected error has occurred.", e)
+        );
+        await();
+    }
+
+    private void loop() {
+        LOG.debug("Started");
+        state = State.RUNNING;
+
+        while (state == State.RUNNING) {
+            long currentTime = Clock.time();
+            Job scheduledJob = scheduledJobs.poll(currentTime);
+            if (scheduledJob != null) {
+                scheduledJob.run();
+            } else {
+                try {
+                    Job job = jobs.poll(scheduledJobs.timeToNext(currentTime), TimeUnit.MILLISECONDS);
+                    if (job != null) {
+                        job.run();
+                    }
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        }
+
+        state = State.STOPPED;
+        LOG.debug("stopped");
+    }
+
+    public static class Clock {
+        private static Supplier<Long> getTime = () -> System.currentTimeMillis();
+        private static Random random = ThreadLocalRandom.current();
+
+        private Clock() {}
+
+        public static void set(Supplier<Long> getTime, Random random) {
```

Comment:
it feels quite weird that you can statically set the `getTime()` arbitrarily?

Reply:
actually everything being static feels a bit odd?

Reply:
Yeah, the idea is the `EventLoop.System` is set once like this:

```
public static void main(...) {
  EventLoop.System.set(clock, randomizer); // set global clock and randomizer that will be used across the systems
  EventLoopGroup elg = new EventLoopGroup(...);
}
```

The main plus side of this style is that we enforce the program to always have one global clock / randomiser. For example, the user won't be able to incorrectly mix and match different kind of clock and randomiser in a single run.

What do you think of this API?
FYI: I've just renamed it from `EventLoop.Clock` since apparently it doesn't only contain clock but also a randomizer

Reply:
The other style which I think is probably nicer is this:

```
EventLoopGroup elg = new EventLoopGroup(...); // use default clock and randomiser
EventLoopGroup elg = new EventLoopGroup(..., clock, randomiser); // supply your own clock and randomiser
```

Reply:
probably prefer constructors rather than setters on statics personally!

Reply:
Great idea. Implemented: https://github.com/graknlabs/common/pull/58/commits/a087d5172c79de8d6f027dfaff765fe245f2e1cc

Reply:
@lriuui0x0 we've actually removed `EventLoop.System` altogether - the user should supply it using the constructor as described above.